x-pack/plugins/apm/server/lib/search_strategies/correlations/async_search_service.ts

```diff
@@ -0,0 +1,264 @@
+    loadedHistogramStepsize: 0,
+    loadedOverallHistogram: 0,
+    loadedFieldCanditates: 0,
+    loadedFieldValuePairs: 0,
+    loadedHistograms: 0,
+    getOverallProgress: () =>
+      progress.loadedHistogramStepsize * 0.025 +
+      progress.loadedOverallHistogram * 0.025 +
+      progress.loadedFieldCanditates * 0.025 +
+      progress.loadedFieldValuePairs * 0.025 +
+      progress.loadedHistograms * 0.9,
+  };
+
+  const values: SearchServiceValue[] = [];
+  let overallHistogram: HistogramItem[] | undefined;
+
+  let percentileThresholdValue: number;
+
+  const cancel = () => {
+    isCancelled = true;
+  };
+
+  const fetchCorrelations = async () => {
+    try {
+      // 95th percentile to be displayed as a marker in the log log chart
+      const percentileThreshold = await fetchTransactionDurationPecentiles(
+        esClient,
+        params,
+        params.percentileThreshold ? [params.percentileThreshold] : undefined
+      );
+      percentileThresholdValue =
+        percentileThreshold[`${params.percentileThreshold}.0`];
+
+      const histogramRangeSteps = await fetchTransactionDurationHistogramRangesteps(
+        esClient,
+        params
+      );
+      progress.loadedHistogramStepsize = 1;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const overallLogHistogramChartData = await fetchTransactionDurationRanges(
+        esClient,
+        params,
+        histogramRangeSteps
+      );
+      progress.loadedOverallHistogram = 1;
+      overallHistogram = overallLogHistogramChartData;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const {
+        fieldCandidates,
+        totalHits,
+      } = await fetchTransactionDurationFieldCandidates(esClient, params);
+
+      progress.loadedFieldCanditates = 1;
+
+      // Create an array of ranges [2, 4, 6, ..., 98]
+      const percents = Array.from(range(2, 100, 2));
+      const percentilesRecords = await fetchTransactionDurationPecentiles(
+        esClient,
+        params,
+        percents
+      );
+      const percentiles = Object.values(percentilesRecords);
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const fieldValuePairs = await fetchTransactionDurationFieldValuePairs(
+        esClient,
+        params,
+        fieldCandidates,
+        progress
+      );
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      async function* fetchTransactionDurationHistograms() {
+        for (const item of shuffle(fieldValuePairs)) {
+          if (item === undefined || isCancelled) {
+            isRunning = false;
+            return;
+          }
+
+          // If one of the fields have an error
+          // We don't want to stop the whole process
+          try {
+            const {
+              correlation,
+              ksTest,
+            } = await fetchTransactionDurationCorrelation(
```

Comment:
I think that recalculating the deduplicated range, percentiles, and expectations for every `fieldValuePairs` is wasteful.

Those things should be calculated once and then passed into this function. The `expectation`, etc. won't change between value pairs as the `percentiles` are not changing (they were gathered before hand).

Reply:
Thanks for catching that! I will address this in the next follow up.